### DISCUSSION TITLE
Get range of parsed input

### DISCUSSION THREAD

**Discussion by dehlen** [pointfreeco/swift-parsing#69] [Q&A] [answered]
Hi,

I am struggling to find a way to get the range of my parsed input and I am not sure whether this is something which would be useful to this library. 

For example consider the following use case: 
I am writing an ingredient parser and I want to make sense of a string by parsing the ingredient amount and unit. But I also want to highlight the parse amount and unit when displaying the ingredient string in a NSAttributedString. From my understand I would need to know which range the parser combinator parsed.

Here is a simple example:

```swift
final class IngredientParser {
    func parse(_ ingredients: [String]) -> [Ingredient] {
        return ingredients.compactMap {
            parseLine($0.trimmingCharacters(in: .whitespacesAndNewlines))
        }
    }

    func parse(_ ingredientsInput: String) -> [Ingredient] {
        return parse(ingredientsInput.split(separator: "\n").map { String($0) })
    }
    
    private func parseLine(_ ingredientsInput: String) -> Ingredient? {        
        let amount: AnyParser<Substring.UTF8View, Int> = Int.parser()
            .eraseToAnyParser()

        let unit: AnyParser<Substring.UTF8View, Unit> = "gram".utf8.map { Unit.gram }.eraseToAnyParser()
        
        let line = Optional.parser(of: amount)
            .skip(Whitespace())
            .take(Optional.parser(of: unit))
            .skip(Whitespace())
            .take(Rest())
            .map { result -> Ingredient in
                return Ingredient(amount: result.0, unit: result.1, name: result.2)
            }
        
        var input = ingredientsInput[...].utf8
        return line.parse(&input)
    }
}
```

Would it be possible to get the parsed range of f.e the Int.parser() from parsing the amount?

**Comment by stephencelis** [comment, marked as answer]
Hi @dehlen,

One way to capture the range of a parsed result is by introducing a new operator that captures it. Here's something @mbrandonw and I sketched out:

```swift
extension Parser where Input: Collection {
  func withRange() -> AnyParser<Input, (output: Output, range: Range<Input.Index>)> {
    .init { input in
      let startIndex = input.startIndex
      guard let output = self.parse(&input)
      else { return nil }
      let endIndex = input.startIndex
      return (output, startIndex ..< endIndex)
    }
  }
}
```

With that defined, you can get the range from `Int.parser()` by tacking on `withRange()`:

```swift
let parser = Int.parser().withRange()
let input = "123 Hello"[...].utf8
let (output, range) = parser.parse(&input)!
XCTAssertEqual(123, output)
XCTAssertEqual(3, input.distance(from: range.lowerBound, to: range.upperBound))
```

Hope this helps!

I'm going to convert this to a discussion, since it's not a bug with the library.

**Reply by dehlen** [reply]
Thank you very much. That is exactly what I was needing and it fits perfectly with the other parser combinators.

**Reply by ZevEisenberg** [reply]
Here is my modernized take on this. I'm rusty with this library, so it could probably be improved. I made a dedicated type instead of a tuple for the `Output`. I did this because I couldn't get the above example to compile when using more than one `withRange()` in the same parser, where the output was a tuple of two or more of the above output/range tuples. I'm not sure why that didn't compile, but this fixes it in my experience:

```swift
struct ValueWithRange<Value, Bound: Comparable> {
  var value: Value
  var range: Range<Bound>
}

struct WithRangeParser<Upstream: Parser, Element>: Parser
  where Upstream.Input: Collection<Element>
{
  typealias Input = Upstream.Input
  typealias Output = ValueWithRange<Upstream.Output, Input.Index>

  let upstream: Upstream

  func parse(_ input: inout Input) throws -> Output {
    let startIndex = input.startIndex
    let output = try upstream.parse(&input)
    let endIndex = input.startIndex
    return ValueWithRange(value: output, range: startIndex..<endIndex)
  }
}

extension Parser where Input: Collection {
  func withRange() -> some Parser<Input, ValueWithRange<Output, Input.Index>> {
    WithRangeParser(upstream: self)
  }
}
```
